framework/include/utils/IndexableProperty.h

```diff
@@ -0,0 +1,154 @@
+InputParameters
+IndexableProperty<T, is_ad>::validParams()
+{
+  auto params = T::validParams();
+  params.template addRequiredParam<MaterialPropertyName>("property",
+                                                         "The name of the material property");
+  params.template addParam<std::vector<unsigned int>>(
+      "component",
+      "Index vector of the scalar component to extract from "
+      "the material property (empty for scalar properties)");
+  return params;
+}
+
+template <typename T, bool is_ad>
+IndexableProperty<T, is_ad>::IndexableProperty(T * host,
+                                               const std::string & property_param,
+                                               const std::string & component_param)
+  : _host(host),
+    _property_param(property_param),
+    _property_name(_host->template getParam<MaterialPropertyName>(_property_param)),
+    _component_param(component_param),
+    _component(host->template getParam<std::vector<unsigned int>>(_component_param))
+{
+  getPropertyHelper<Real>(_property_real, 0);
+  getPropertyHelper<RealVectorValue>(_property_real_vector_value, 1);
+  getPropertyHelper<std::vector<Real>>(_property_std_vector, 1);
+  getPropertyHelper<RankTwoTensor>(_property_rank_two_tensor, 2);
+  getPropertyHelper<RankThreeTensor>(_property_rank_three_tensor, 3);
+  getPropertyHelper<RankFourTensor>(_property_rank_four_tensor, 4);
+
+  if (!*this)
+    mooseError("The ",
+               is_ad ? "AD" : "non-AD",
+               " material property '",
+               _property_name,
+               "' does not exist");
+}
+
+template <typename T, bool is_ad>
+GenericReal<is_ad>
+IndexableProperty<T, is_ad>::operator[](int qp) const
+{
+  if (_property_real)
+    return (*_property_real)[qp];
+  if (_property_std_vector)
+    return (*_property_std_vector)[qp][_component[0]];
+  if (_property_real_vector_value)
+    return (*_property_real_vector_value)[qp](_component[0]);
+  if (_property_rank_two_tensor)
+    return (*_property_rank_two_tensor)[qp](_component[0], _component[1]);
+  if (_property_rank_three_tensor)
+    return (*_property_rank_three_tensor)[qp](_component[0], _component[1], _component[2]);
+  if (_property_rank_four_tensor)
+    return (*_property_rank_four_tensor)[qp](
+        _component[0], _component[1], _component[2], _component[3]);
+  mooseError("internal error");
+}
+
+template <typename T, bool is_ad>
+IndexableProperty<T, is_ad>::operator bool() const
+{
+  return _property_real || _property_std_vector || _property_real_vector_value ||
+         _property_rank_two_tensor || _property_rank_three_tensor || _property_rank_four_tensor;
+}
+
+template <typename T, bool is_ad>
+template <typename P>
+void
+IndexableProperty<T, is_ad>::getPropertyHelper(const GenericMaterialProperty<P, is_ad> *& pointer,
+                                               unsigned int components)
+{
+  if (_host->template hasGenericMaterialProperty<P, is_ad>(_property_param))
```

Comment:
does it make more sense to have the material property interface at the host level or at the IndexableProperty? I feel like the first one is more convoluted but probably more convenient?

Reply:
the second one doesnt work I think.

I think what I would like best would be to see this code work within the MaterialPropertyInterface. Maybe there's a hold up there I have not thought things through as deeply

Reply:
let's merge this as is now and if it turns out to be of greater interest we can always migrate it into the mat prop interface

Reply:
I'm ok with that if you need this object now.  But I'm not a fan of this interface and if we deprecate it soon then should we add a note to the documentation or source to advise developers to contact us if they want to use this?
What has been done for less ideal merges before?